I think the overflow warning has its origin in the default integer size for a given architecture. If for example the type of a literal is inferred to be `Word 32` in Copilot (and the value fits in 32 bits), but the compiler for the target architecture defaults to something smaller, e.g. a `uint16_t`, an overflow can occur that will not be detected by Copilot. The C compiler can in this case come up with the warning. Using your changes to Copilot should indeed fix this problem.

Strangely enough, compiling the following piece of testcode
```c
uint16_t v16 = 1 / 65536; /* 2^16 */
uint32_t v32 = 1 / 4294967296; /* 2^32 */
uint64_t v64 = 1 / 18446744073709551616; /* 2^64 */
```
With `avr-gcc -Wall -Wpedantic -std=c99 -mmcu=atmega328p` (to enable the microprocessor which the Arduino Uno is based on), gives us:
```c
main.c:17:22: warning: integer constant is too large for its type
   17 |   uint64_t v64 = 1 / 18446744073709551616;
      |                      ^~~~~~~~~~~~~~~~~~~~
main.c:17:20: warning: division by zero [-Wdiv-by-zero]
   17 |   uint64_t v64 = 1 / 18446744073709551616;
```
(note that the div-by-zero is thrown, because 2^64 obviously overflows to 0)
So I guess the literal integers default to 64 bits, even on the 8-bit ATmega 328p? This seems unlikely, and judging from your output, isn't the case with the Arduino IDE.

Actually if we try to write 9223372036854775808 we get the same problem, 9223372036854775807 (one less) works, so we can conclude that literals are signed by default.

Integers on the ATmega AVR architecture are always two bytes, so I would expect 2^15 to be already problematic.

Is it possible for you to look up the exact arguments the IDE uses to compile your code (I tried to find it online, but couldn't)? Maybe there are some parameters to GCC that changes the size of the literals?

https://gcc.gnu.org/onlinedocs/gcc/AVR-Options.html shows there is an option `-mint8` which makes `int` always and `int8` (this even removes uint64_t). Enabling this will make it start complain at 2147483648 (2^31 - 1)

For now, I will leave this pull request open, as I would like to get clear what exactly is going on with the default int size. I will however merge the request afterwards.

--- a/src/Copilot/Compile/C99/Translate.hs
+++ b/src/Copilot/Compile/C99/Translate.hs
@@ -112,16 +112,16 @@ transop3 op e1 e2 e3 = case op of
 constty :: Type a -> a -> C.Expr
 constty ty = case ty of
   Bool   -> C.LitBool
-  Int8   -> explicit_type ty . C.LitInt . fromIntegral
-  Int16  -> explicit_type ty . C.LitInt . fromIntegral
-  Int32  -> explicit_type ty . C.LitInt . fromIntegral
-  Int64  -> explicit_type ty . C.LitInt . fromIntegral
-  Word8  -> explicit_type ty . C.LitInt . fromIntegral
-  Word16 -> explicit_type ty . C.LitInt . fromIntegral
-  Word32 -> explicit_type ty . C.LitInt . fromIntegral
-  Word64 -> explicit_type ty . C.LitInt . fromIntegral
-  Float  -> explicit_type ty . C.LitFloat
-  Double -> explicit_type ty . C.LitDouble
+  Int8   -> explicitty ty . C.LitInt . fromIntegral
+  Int16  -> explicitty ty . C.LitInt . fromIntegral
+  Int32  -> explicitty ty . C.LitInt . fromIntegral
+  Int64  -> explicitty ty . C.LitInt . fromIntegral
+  Word8  -> explicitty ty . C.LitInt . fromIntegral
+  Word16 -> explicitty ty . C.LitInt . fromIntegral
+  Word32 -> explicitty ty . C.LitInt . fromIntegral
+  Word64 -> explicitty ty . C.LitInt . fromIntegral
+  Float  -> explicitty ty . C.LitFloat
+  Double -> explicitty ty . C.LitDouble
   Struct _ -> \v -> C.InitVal (transtypename ty) (map fieldinit (toValues v))
     where
       fieldinit (Value ty (Field val)) = C.InitExpr $ constty ty val
@@ -129,8 +129,8 @@ constty ty = case ty of
     where
       valinit = C.InitExpr . constty ty'
 
-explicit_type :: Type a -> C.Expr -> C.Expr
-explicit_type ty = C.Cast (transtypename ty)
+explicitty :: Type a -> C.Expr -> C.Expr
+explicitty ty = C.Cast (transtypename ty)
 
 -- | Translate a Copilot type to a C99 type.
 transtype :: Type a -> C.Type

